### PR TITLE
New: Add class-methods-use-this rule

### PR DESCRIPTION
**Pull request by pwmckenna**
https://github.com/eslint/eslint/issues/5139

### CODE REVIEW

**Comment by pwmckenna** [comment]
just signed the CLA.

File: tests/lib/rules/class-methods-use-this.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ruleTester.run("class-methods-use-this", rule, {
+    valid: [
+        {code: "class A { constructor() {} }", parserOptions: { ecmaVersion: 6 }},
+        {code: "class A { foo() {this} }", parserOptions: { ecmaVersion: 6 }},
+        {code: "class A { foo() {this.bar = 'bar';} }", parserOptions: { ecmaVersion: 6 }},
+        {code: "class A { foo() {bar(this);} }", parserOptions: { ecmaVersion: 6 }},
+        {code: "class A { foo() { if(true) { return this; } } }", parserOptions: { ecmaVersion: 6 }},
+        {code: "class A { static foo() {} }", parserOptions: { ecmaVersion: 6 }},
+        {code: "class A { constructor() {} }", parserOptions: { ecmaVersion: 6 }}
+    ],
```
Comment (michaelficarra) [inline tests/lib/rules/class-methods-use-this.js]: Add this test:

```js
({ a(){} });
```

**Comment by michaelficarra** [comment]
You need to fix your commit message. See the Travis failure.

**Comment by pwmckenna** [comment]
@michaelficarra fixed up the commit message. thanks for the pointer.

File: lib/rules/class-methods-use-this.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+     */
+    function markThisUsed() {
+        if (stack.length) {
+            stack[stack.length - 1] = true;
+        }
+    }
+
+    return {
+        "FunctionDeclaration": enterFunction,
+        "FunctionDeclaration:exit": exitFunction,
```
Comment (michaelficarra) [inline lib/rules/class-methods-use-this.js]: These shouldn't be necessary.

File: tests/lib/rules/class-methods-use-this.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+//------------------------------------------------------------------------------
+// Tests
+//------------------------------------------------------------------------------
+
+var ruleTester = new RuleTester();
+ruleTester.run("class-methods-use-this", rule, {
+    valid: [
+        {code: "class A { constructor() {} }", parserOptions: { ecmaVersion: 6 }},
+        {code: "class A { foo() {this} }", parserOptions: { ecmaVersion: 6 }},
```
Comment (michaelficarra) [inline tests/lib/rules/class-methods-use-this.js]: I would add

```js
class A { foo() { () => this; } }
```

**Comment by michaelficarra** [comment]
LGTM.

File: tests/lib/rules/class-methods-use-this.js (hunk 3 above)
Comment (pwmckenna) [inline tests/lib/rules/class-methods-use-this.js]: just added this test, and the corresponding invalid version...which honestly is baffling to me.

Here's the invalid test:
`class A { foo() { () => undefined; } }`

This fails the way I would want it to, but not the way I *expected* it to. `=>` binds this, so at some point `this` is being used no? Anyhow, I'm happy with the additional tests, even if I don't fully understand them.

Comment (michaelficarra) [inline tests/lib/rules/class-methods-use-this.js]: Nope, it's only used if it's actually mentioned.

**Comment by gyandeeps** [comment]
LGTM

**Comment by ilyavolodin** [comment]
Can you update commit message to say `(fixes #5139)` instead of `refs`? Also, I'm not crazy about the name of this rule. But can't really come up with anything else. @eslint/eslint-team any suggestions?

**Comment by platinumazure** [comment]
What about something like `prefer-static-methods` or `prefer-static`?

**Comment by ilyavolodin** [comment]
@platinumazure I think this is the rule for the opposite. Something like `no-class-static-methods` is probably more descriptive, but not any shorter, or better then what's already there.

**Comment by mysticatea** [comment]
In my understanding, this rule is just disallowing prototype methods which don't use `this`/`super`. Those can be moved to static methods or functions. And substitutes are not only static methods (I would often use module scoped functions).

Hmm, `class-methods-use-this`, `use-this-in-methods`, ...

**Comment by platinumazure** [comment]
How about no-unnecessary-instance-methods then? Unfortunately it's very
long, but I don't think you can do much better on clarity.

**Comment by mysticatea** [comment]
It's nice to me, but long :expressionless:

**Comment by michaelficarra** [comment]
How about `class-methods-use-this`?

**Comment by mysticatea** [comment]
No problem to me.

File: docs/rules/class-methods-use-this.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
+# Enforce that class methods utilize `this`. (class-methods-use-this)
```
Comment (nzakas) [inline docs/rules/class-methods-use-this.md]: No period at end of sentence.

File: docs/rules/class-methods-use-this.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+If a class method does not use this, it can safely be made a static function.
```
Comment (nzakas) [inline docs/rules/class-methods-use-this.md]: Need a lot more in this section. Include some examples and explain what you're talking about. See http://eslint.org/docs/rules/no-undefined as an example.

File: docs/rules/class-methods-use-this.md (hunk 4 above)
Comment (nzakas) [inline docs/rules/class-methods-use-this.md]: Oops sorry, I meant please remove the period. Also, capitalize the words.

**Comment by nzakas** [comment]
@pwmckenna are you still working on this?

**Comment by nzakas** [comment]
@pwmckenna just following up one last time: do you intend to keep working on this?

**Comment by nzakas** [comment]
No response, so closing.

**Pull request closed without merging**